Create a cumulative displacement design plan and frame metadata tile generation utility (#2)
ASFHyP3/OPERA-DISP-TMS: Design.md

These 1x1 degree COGs are then summarized into a set of files that are ingestable into a [MapBox tileset source](https://docs.mapbox.com/mapbox-tiling-service/guides/tileset-sources/), the core unit of the MapBox Tiling Service. The requirements for this are as follows:
- UInt8 datatype
- Total dataset size of less than 50 GB
- No file greater than 10 GB
- No more than 10 input files per tileset source
A single tile map layer can be composed of multiple tileset sources if necessary to get around the 10 file limit, but testing has shown that we can meet the requirements above for a single tileset source that includes all of North America, which is sufficient for this project.

> To go from the 1x1 degree COGs to a MapBox tileset source we must mosaic the 1x1 degree COGs into 10 or fewer tiled GeoTiffs with a UInt8 datatype. GeoTiffs are used because the overviews included in COGs are unused by MapBox, and only serve to increase file size. Dataset-specific discussions will need to had with the OPERA team concerning how to represent the datasets as UInt8 datatypes, but an initial guess of `round(pixel_value,2)*100) + 1` will be used for all datsets during initial development. Values above and below this range will be clipped to 255 and 0 respectively.

Reviewer: ```suggestion
To go from the 1x1 degree COGs to a MapBox tileset source, we must mosaic the 1x1 degree COGs into 10 or fewer tiled GeoTiffs with a UInt8 datatype. GeoTiffs are used because the overviews included in COGs are unused by MapBox and only serve to increase file size. Dataset-specific discussions will need to be had with the OPERA team concerning how to represent the datasets as UInt8 datatypes, but an initial guess of `round(pixel_value,2)*100) + 1` will be used for all datasets during initial development. Values above and below this range will be clipped to 255 and 0, respectively.
```

LZW or another compression method will be used for all files to reduce file sizes.

## Cumulative Displacement
### Format

> To fully understand the context of InSAR-derived displacement data (such as the OPERA-Disp products) multiple pieces are required for each pixel:

Reviewer: ```suggestion
To fully understand the context of InSAR-derived displacement data (such as the OPERA-Disp products), multiple pieces are required for each pixel:
```

1. A UInt16 band indicating the frame number the cumulative displacement value each pixel is sourced from
2. Geotiff metadata attributes detailing the reference date and reference point that will be enforced for all tiles created using the metadata raster.

For example, the geotiff metadata attributes may look like:
```
OPERA_FRAMES: "1, 2"
FRAME_1_REF_POINT: "129877.23, 128383.28"
FRAME_1_EPSG: "12345"
FRAME_1_REF_POINT_ARRAY: "10, 20"
FRAME_1_REF_DATE: "2015/01/01"
FRAME_2_REF_POINT: "22348.23, 38973.28"
FRAME_2_EPSG: "12345"
FRAME_2_REF_POINT_ARRAY: "20, 11"
FRAME_2_REF_DATE: "2015/02/01"
```
This is similar to the short wavelength displacement metadata fields, but with the exclusion of the secondary date and metadata raster URL information.

**Note: a time-series stack of OPERA-DISP products from the same frame are not guaranteed to have the same reference date or location!**

> However having the same spatiotemporal reference for each pixel in a time-series stack is an important requirement for date/location re-referencing and velocity calcultions. Thus, we standardize the spatiotemporal reference for each product before mosaicking. 

Reviewer: ```suggestion
However, having the same spatiotemporal reference for each pixel in a time-series stack is an important requirement for date/location re-referencing and velocity calculations. Thus, we standardize the spatiotemporal reference for each product before mosaicking. 
```

- Specify a date range you want to generate a tile for
- Generate a new raster with the same projection, resolution, and extent as the metadata raster that contains short_wavelength_displacement data
- Obtain the list of needed frames from the metadata raster metadata tags
- For each needed frame
    - Find OPERA-DISP product with the latest available secondary date within the date range
        - If a frame has no available product within the date range, set its pixels to nodata
    - Check that the product's reference date and reference point are the same as the metadata raster
        - If this is not the case, re-reference the product (methodology will be written soon)
    - Reproject the product's cumulative displacement band to the same projection, resolution, and extent as the metadata raster
    - For the cells that correspond to the frame in question within the new raster
        - Set the cumulative displacement band values to the ones from the reprojected OPERA-DISP product
- Copy the spatiotemporal reference metadata from the metadata raster to the new raster
- Write to COG with LZW compression

### MapBox Tileset Generation

> TODO

Author: No, this section won't be completed in this PR.

# Mosaic Design
## General Concept
There are two key steps for preparing OPERA DISP data for presentation in the MapBox Tiling Service:

1. The discrete, non-rectangular, and rotated frames of OPERA DISP products need to be summarized into a continuous single-layer raster representation.
2. This continuous representation needs to delivered to MapBox in a format that is simple for it to ingest.

> To meet the first requirement, we take discrete OPERA DISP products and transform them into a set of square 1x1 degree COGs. This strategy has worked well for other projects (such as the Global Seasonal Coherence dataset) and allows us to easily convert the data to the tileset representation that MapBox. It also provides us with a great starting point to provide this data to users via other avenues (GIBS, ESRI Image Server, TMS, etc.) in the future. There are intricacies for mosaicking each product, which are detailed in other sections, but in general each tile is created separately using 1-4 input OPERA DISP products. Products are overlaid so that products along the same absolute orbit are continuous, and neighboring relative orbit data are overlaid from east to west. The 1x1 degree COGs maintain the same datatype and data values as the original products.

Reviewer 2: Generating the 1x1 data set sounds like an opportunity rather than a requirement, yes? Given the time constraints of the project, we could focus on building the "10 or fewer tiled GeoTiffs with a UInt8 datatype", rather than adding the requirement of the intermediate 1x1 COG data set.

These 1x1 degree COGs are then summarized into a set of files that are ingestable into a [MapBox tileset source](https://docs.mapbox.com/mapbox-tiling-service/guides/tileset-sources/), the core unit of the MapBox Tiling Service. The requirements for this are as follows:
- UInt8 datatype
- Total dataset size of less than 50 GB
- No file greater than 10 GB
- No more than 10 input files per tileset source
A single tile map layer can be composed of multiple tileset sources if necessary to get around the 10 file limit, but testing has shown that we can meet the requirements above for a single tileset source that includes all of North America, which is sufficient for this project.

> To go from the 1x1 degree COGs to a MapBox tileset source we must mosaic the 1x1 degree COGs into 10 or fewer tiled GeoTiffs with a UInt8 datatype. GeoTiffs are used because the overviews included in COGs are unused by MapBox, and only serve to increase file size. Dataset-specific discussions will need to had with the OPERA team concerning how to represent the datasets as UInt16 datatypes, but an initial guess of `round(pixel_value,2)*100) + 1` will be used for all datsets during initial development. Values above and below this range will be clipped to 255 and 0 respectively.

Reviewer 2: > The requirements for this are as follows:
>   - UInt8 datatype

> ...how to represent the datasets as UInt16 datatypes

Is one of these a typo?

LZW or another compression method will be used for all files to reduce file sizes.

## Cumulative Displacement
### Format
To fully understand the context of InSAR-derived displacement data (such as the OPERA-Disp products) multiple pieces are required for each pixel:

1. The geographic location of the pixel (from the GDAL metadata)
2. The start and end date for the displacement observation
3. The reference pixel location to which all displacement measurements in a product are relative to

This is a lot of information, and we will need to ensure that we provide access to all of this data (or at least make it trackable) for every pixel in the mosaic.

To do this, the 1x1 degree cumulative displacement COGs will have the following components:

> 1. A Float32 short wavelength cumulative displacement band sourced directly from the products

Reviewer 2: as of v0.4, disp products include:
```
displacement (32-bit floating-point)
short_wavelength_displacement (64-bit floating-point)
connected_component_labels (16-bit unsigned integer)
temporal_coherence (32-bit floating-point)
interferometric_correlation (32-bit floating-point)
persistent_scatterer_mask (8-bit unsigned integer)
unwrapper_mask (8-bit unsigned integer)
/corrections/tropospheric_delay (32-bit floating-point)
/corrections/ionospheric_delay (32-bit floating-point)
/corrections/solid_earth_tide (32-bit floating-point)
/corrections/plate_motion (32-bit floating-point)
```

Are we planning to use `displacement` or `short_wavelength_displacement`?

> 2. A sidecar metadata file (likely a STAC item) containing frame, spatiotemporal reference data, and a link to an external tile frame image (see below)

Reviewer 2: > A sidecar metadata file (likely a STAC item)

I'd recommend saving these attributes as key/value metadata attributes of the GeoTIFF itself, rather than in a sidecar file.

The sidecar metadata will contain the following fields:
1. List of OPERA frames present in the cumulative displacement COG
2. The frame reference and secondary date in the format YYYY/MM/DD for each frame present in the tile
3. The frame reference point location encoded as both the array coordinates (row/column) and the geographic coordinates of the reference pixel in the native projection of the frame for each frame present in the tile
4. A URL pointing to the location of OPERA frame image for that tile (see below)

For example, the fields in metadata raster may look like:
```
OPERA_FRAMES: "1, 2"
FRAME_1_REF_POINT: "129877.23, 128383.28"
FRAME_1_EPSG: "12345"
FRAME_1_REF_POINT_ARRAY: "10, 20"
FRAME_1_REF_DATE: "2015/01/01"

> FRAME_1_SEC_DATE: "2015/01/02"

Reviewer 2: ```suggestion
FRAME_1_SEC_DATE: "2015/01/02"
FRAME_1_METADATA_URL: "https://mybucket.s3-us-west-2.amazonaws.com/metadata_tiffs/1.tiff"
```

FRAME_2_REF_POINT: "22348.23, 38973.28"
FRAME_2_EPSG: "12345"
FRAME_2_REF_POINT_ARRAY: "20, 11"
FRAME_2_REF_DATE: "2015/02/01"

> FRAME_2_SEC_DATE: "2015/02/02"

Reviewer 2: ```suggestion
FRAME_2_SEC_DATE: "2015/02/02"
FRAME_2_METADATA_URL: "https://mybucket.s3-us-west-2.amazonaws.com/metadata_tiffs/2.tiff"
```

```

In addition to this per-COG information, there will also be one accompanying metadata raster (in COG format) for each tile location (i.e., one per footprint). The metadata raster will have the same resolution, extent, and projection as the cumulative displacement COGs, and in fact will serve as the template for create cumulative displacement COGs. The metadata raster will have the following components
1. A UInt16 band indicating the frame number the cumulative displacement value each pixel is sourced from

> 2. A sidecar metadata file with the same information as the cumulative displacement sidecar files, except for secondary date and OPERA frame image URL

Reviewer 2: Reference date, secondary date, and reference point are all specific to a particular disp .nc product, and might all vary from product to product from the same frame, yes? What's the intent of capturing a reference point and reference date for the frame? Are we using this to document the fixed references all 1x1 COGs for that frame will be re-referenced to when we generate them?

**Note: a time-series stack of OPERA-DISP products from the same frame are not guaranteed to have the same reference date or location!**

However having the same spatiotemporal reference for each pixel in a time-series stack is an important requirement for date/location re-referencing and velocity calcultions. Thus, we standardize the spatiotemporal reference for each product before mosaicking. 

Note that we only standardize the spatiotemporal reference so that each location is consistent through time. We do not standardize this information across space. Non-continuous frame coverage, and varying data collection dates makes standardizing across space a difficult challenge that is planned to be undertaken during the creation of the OPERA-DISP Vertical Land Motion project.

### 1x1 Degree Tile Creation
Prior to mosaicking of OPERA-DISP products, we generate the metadata rasters.

> For each 1x1 degree tile, there is one metadata raster for each orbit direction, for a total of two per tile. The metadata rasters will each span a 1x1 degree area in the web mercator (EPSG:3857) projection with a 30 m pixel spacing. The frame band is created by cross-referencing each tile's extent with the OPERA-DISP frame footprints. A geospatial representation of the OPERA-DISP frame footprints can be created using the [OPERA burst_db utility](https://github.com/opera-adt/burst_db). Frames will be overlaid so that frames from the same relative orbit are contiguous. Relative orbits are stacked from East to West on the ascending pass (West on top), and from West to East on the descending pass (East on top). Within a relative orbit, frames are overlaid so that older frames are on top of younger frames. This corresponds to North on top for the ascending pass, and South on top for the descending pass.

Reviewer 2: > Relative orbits are stacked from East to West on the ascending pass (West on top), and from West to East on the descending pass (East on top)

Using "biggest frame number on top" feels simpler to implement, and might give an equally pleasing mosaic. Along-track frames have consecutive frame ids.

Here's a rendering in qgis of what layering "biggest on top" would look like:
![Screenshot from 2024-09-23 09-43-31](https://github.com/user-attachments/assets/2a75027a-7ce9-413f-a941-8cef9496b57e)

Once the frame band has been constructed, we set the spatiotemporal reference metadata using the values present in the earliest product available for each frame.

We then use the generated metadata rasters as a guide for creating all 1x1 degree cumulative displacement tiles in a time-series stack. The procedure is as follows:

- Specify a date range you want to generate a tile for
- Generate a new raster with the same projection, resolution, and extent as the metadata raster with three bands
    - Cumulative displacement

>     - Days since 2014/01/01

Reviewer 2: The intent here is to document the secondary date for each pixel and "days since" let's us express that as an int (a valid datatype for a geotiff band), rather than a date or string?

If so, Unix timestamp feels like a more standardized way of representing a date as an integer.

Reviewer 2: Do we need to document reference data per pixel as well?

>     - Frame number

Reviewer 2: What value does the frame number band add? This would be an exact duplicate of the frame number band of the metadata tiff, yes?

- Obtain the list of needed frames from the metadata raster metadata tags
- For each needed frame
    - Find OPERA-DISP product with the latest available secondary date within the date range
        - If a frame has no available product within the date range, raise an exception
    - Check that the product's spatiotemporal reference date is the same as the metadata raster
        - If this is not the case, re-reference the product (methodology will be written soon)
    - Reproject the product's cumulative displacement band to the same projection, resolution, and extent as the metadata raster
    - For the cells that correspond to the frame in question within the new raster
        - Set the cumulative displacement band values to the ones from the reprojected OPERA-DISP product

>         - Set the Days since 2014/01/01 to the secondary date of the OPERA-DISP product

Reviewer 2: ```suggestion
        - Set the Days since 2014/01/01 based on the secondary date of the OPERA-DISP product
```
